src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java

```diff
@@ -526,6 +544,122 @@ public synchronized void deprovision(Localizable reason) throws Exception {
                 new Object[]{this.getDisplayName(), reason == null ? "Unknown reason" : reason.toString()});
 
         computer.setAcceptingTasks(false);
+
+        ComputerLauncher launcher = computer.getLauncher();
+
+        if ((launcher != null && launcher instanceof AzureVMAgentSSHLauncher)) {
+            AzureVMAgentSSHLauncher azureLauncher = (AzureVMAgentSSHLauncher) launcher;
+
+            OutputStream terminateResults = new OutputStream() {
+                private String terminateResultBuffer = "";
+
+                @Override
+                public void write(int b) throws IOException {
+                    terminateResultBuffer += (char) b;
+                }
+
+                @Override
+                public void flush() {
+                    if (!terminateResultBuffer.isEmpty()) {
+                        LOGGER.log(Level.FINER, "AzureVMAgent: deprovision: terminateScript, " + terminateResultBuffer);
+                        terminateResultBuffer = "";
+                    }
+                }
+            };
+
+            PrintStream terminateStream = new PrintStream(terminateResults, true, "UTF-8");
+
+            final boolean isUnix = this.getOsType().equals(OperatingSystemTypes.LINUX);
+            // Check if VM is already stopped or stopping or getting deleted ,
+            // if yes then there is no point in trying to connect
+            // Added this check - since after restarting jenkins master,
+            // jenkins is trying to connect to all the agents although agents are suspended.
+            // This still means that a delete agent will eventually get cleaned up.
+            try {
+                if (!this.isVMAliveOrHealthy()) {
+                    LOGGER.log(Level.INFO,
```

Comment:
Only one log in this block?

Reply:
It is all mostly initialization and then when the actual work is being done during the termination script execution the loggingThread  handles pushing the relevant information out.

Reply:
These codes seem to be copied from `AzureVMAgentSSHLauncher#launch(SlaveComputer agentComputer, TaskListener listener)` method. And in that method, when the vm is not alive or healthy, it will directly return and do not try to launch the computer. But this check here did nothing. Do you want to make sure the vm is alive and healthy and then execute the terminate script?

Reply:
It actually is because of some obfusticated functionality with the isVMAliveOrHealthy.  Since that function will loop until the initial connection is completed, in the event of a Jenkins restart (but before the deprovision criteria is met) this ensures that the terminate script will not be executed until the SSH connection has been re-established.  Previously deprovision could be called and processed without ever having the SSH connection establish because there was no need to if the VM was to be deprovisioned.  Now with the terminate script logic it needs to actually let the SSH establish itself to let the script execute before the other tear down processes are run.